File: generate-egifts.js
const fs = require("fs");

// Configuration
const TOTAL_RECORDS = 1000;
const OUTPUT_FILE = "egift.txt";

// Default value for all egifts
const DEFAULT_VALUE = 150;

// Date range: 05/11/2025 to 09/11/2025
const START_DATE = new Date("2025-11-05");
const END_DATE = new Date("2025-11-09");

// Helper function to generate random alphanumeric string
function generateRandomString(length) {
  const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
  let result = "";
  for (let i = 0; i < length; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}

// Helper function to generate random date between start and end
function randomDate(start, end) {
  const timestamp =
    start.getTime() + Math.random() * (end.getTime() - start.getTime());
  return new Date(timestamp);
}

// Generate egift records
function generateEgifts() {
  const records = [];

  console.log(`Generating ${TOTAL_RECORDS} egift records...`);

  for (let i = 1; i <= TOTAL_RECORDS; i++) {
    // Generate completely random code: XXXX-XXXXXXXX-XXXX
    const code = `${generateRandomString(4)}-${generateRandomString(
      8
    )}-${generateRandomString(4)}`;

    // Generate 4-digit PIN
    const pin = Math.floor(1000 + Math.random() * 9000).toString();

    // Use default value
    const value = DEFAULT_VALUE;

    // Selling price: 75% to 90% of value
    const discountPercent = 0.75 + Math.random() * 0.15;
    const sellingPrice = Math.floor(value * discountPercent);

    // Random purchase date
    const purchaseDate = randomDate(START_DATE, END_DATE);
    const dateStr = purchaseDate.toISOString().split("T")[0];

    // Add record
    records.push({
      id: i,
      code: code,
      pin: pin,
      value: value,
      sellingPrice: sellingPrice,
      purchaseDate: dateStr,
    });

    // Progress indicator
    if (i % 100 === 0) {
      console.log(`  ${i}/${TOTAL_RECORDS} records generated...`);
    }
  }

  return records;
}

// Write to file
function writeToFile(records) {
  console.log(`\nWriting to ${OUTPUT_FILE}...`);

  // Create CSV header
  const header = "ID,Code,PIN,Value,SellingPrice,PurchaseDate";

  // Convert records to CSV lines
  const lines = records.map((record) => {
    return `${record.id},${record.code},${record.pin},${record.value},${record.sellingPrice},${record.purchaseDate}`;
  });

  // Combine header and lines
  const content = [header, ...lines].join("\n");

  // Write to file
  fs.writeFileSync(OUTPUT_FILE, content, "utf8");

  console.log(
    `✅ Successfully created ${OUTPUT_FILE} with ${records.length} records!`
  );
  console.log(
    `\nFile size: ${(Buffer.byteLength(content) / 1024).toFixed(2)} KB`
  );
}

// Statistics
function showStatistics(records) {
  console.log("\n📊 Statistics:");
  console.log(`  Total records: ${records.length}`);
  console.log(`  All records have value: $${DEFAULT_VALUE}`);

  // Date distribution
  const dateCount = {};
  records.forEach((r) => {
    dateCount[r.purchaseDate] = (dateCount[r.purchaseDate] || 0) + 1;
  });

  console.log("\n  Records by purchase date:");
  Object.entries(dateCount)
    .sort()
    .forEach(([date, count]) => {
      console.log(`    ${date}: ${count}`);
    });

  // Selling price range
  const prices = records.map((r) => r.sellingPrice);
  const minPrice = Math.min(...prices);
  const maxPrice = Math.max(...prices);
  const avgPrice = Math.floor(
    prices.reduce((a, b) => a + b, 0) / prices.length
  );

  console.log("\n  Selling price range:");
  console.log(`    Min: $${minPrice}`);
  console.log(`    Max: $${maxPrice}`);
  console.log(`    Average: $${avgPrice}`);
}

// Main execution
console.log("🎁 E-Gift Card Generator\n");
console.log("Configuration:");
console.log(`  Records: ${TOTAL_RECORDS}`);
console.log(
  `  Date range: ${START_DATE.toISOString().split("T")[0]} to ${
    END_DATE.toISOString().split("T")[0]
  }`
);
console.log(`  Default Value: $${DEFAULT_VALUE}`);
console.log(`  Code Format: XXXX-XXXXXXXX-XXXX (completely random)\n`);

// Generate records
const records = generateEgifts();

// Write to file
writeToFile(records);

// Show statistics
showStatistics(records);

console.log("\n✨ Done!\n");

File: generate_egifts.py
import random
import string
import argparse
from datetime import datetime, timedelta

def generate_code():
    """Generate a code in format XXXX-XXXXXXXX-XXXX"""
    chars = string.ascii_uppercase + string.digits
    part1 = ''.join(random.choices(chars, k=4))
    part2 = ''.join(random.choices(chars, k=8))
    part3 = ''.join(random.choices(chars, k=4))
    return f"{part1}-{part2}-{part3}"

def generate_pin():
    """Generate a 4-digit PIN"""
    return random.randint(1000, 9999)

def generate_selling_price(value):
    """Generate selling price based on value"""
    ratio = random.uniform(0.75, 0.90)
    return round(value * ratio)

def generate_random_date():
    """Generate random date between 2026-01-03 and 2026-01-12"""
    start_date = datetime(2026, 1, 3)
    end_date = datetime(2026, 1, 12)
    delta = (end_date - start_date).days
    random_days = random.randint(0, delta)
    return (start_date + timedelta(days=random_days)).strftime('%Y-%m-%d')

def generate_egifts_by_amount(total_amount, output_file='new_egifts.csv'):
    """Generate egift entries based on total amount"""
    values = [150, 300, 500]
    egifts = []
    remaining = total_amount
    
    while remaining > 0:
        available_values = [v for v in values if v <= remaining]
        if not available_values:
            break
        value = random.choice(available_values)
        egifts.append(value)
        remaining -= value
    
    with open(output_file, 'w', encoding='utf-8', newline='') as f:
        # Write header
        f.write("ID,Code,PIN,Value,SellingPrice,PurchaseDate\n")
        
        for value in egifts:
            code = generate_code()
            pin = generate_pin()
            selling_price = generate_selling_price(value)
            purchase_date = generate_random_date()
            
            # No ID as per user request
            f.write(f",{code},{pin},{value},{selling_price},{purchase_date}\n")
    
    actual_total = sum(egifts)
    print(f"Successfully generated {len(egifts)} egifts (total: {actual_total}) in {output_file}")

def generate_egifts(count=360, output_file='new_egifts.csv'):
    """Generate egift entries"""
    values = [150, 300, 500]
    
    with open(output_file, 'w', encoding='utf-8', newline='') as f:
        # Write header
        f.write("ID,Code,PIN,Value,SellingPrice,PurchaseDate\n")
        
        for i in range(count):
            code = generate_code()
            pin = generate_pin()
            value = random.choice(values)
            selling_price = generate_selling_price(value)
            purchase_date = generate_random_date()
            
            # No ID as per user request
            f.write(f",{code},{pin},{value},{selling_price},{purchase_date}\n")
    
    print(f"Successfully generated {count} egifts in {output_file}")

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Generate eGift entries')
    parser.add_argument('-n', '--count', type=int, default=360, help='Number of egifts to generate')
    parser.add_argument('-o', '--output', type=str, default='new_egifts.csv', help='Output file name')
    parser.add_argument('-a', '--amount', type=int, help='Total amount of egifts (overrides count)')
    
    args = parser.parse_args()
    
    if args.amount:
        generate_egifts_by_amount(args.amount, args.output)
    else:
        generate_egifts(args.count, args.output)
